File: services/price-service/internal/storage/local.go
package storage

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// LocalStorage implements Storage interface using local filesystem
type LocalStorage struct {
	basePath string
}

// NewLocalStorage creates a new local filesystem storage
func NewLocalStorage(basePath string) (*LocalStorage, error) {
	// Ensure base path exists
	if err := os.MkdirAll(basePath, 0755); err != nil {
		return nil, fmt.Errorf("failed to create storage directory %s: %w", basePath, err)
	}

	return &LocalStorage{
		basePath: basePath,
	}, nil
}

const MinCompressionSize = 1024 // 1KB minimum

// Put stores content at the given key with optional metadata
func (s *LocalStorage) Put(ctx context.Context, key string, content []byte, metadata *Metadata) error {
	fullPath := s.keyToPath(key)

	// Ensure parent directory exists
	dir := filepath.Dir(fullPath)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return fmt.Errorf("failed to create directory %s: %w", dir, err)
	}

	var dataToStore []byte
	var compressed bool
	storePath := fullPath

	// Determine if we should compress
	if metadata != nil && len(content) >= MinCompressionSize {
		fileType := ""
		if metadata.Custom != nil {
			fileType = metadata.Custom["file_type"]
		}
		if ShouldCompress(fileType) {
			compressedData, err := CompressWithGzip(content)
			if err == nil && len(compressedData) < len(content) {
				dataToStore = compressedData
				compressed = true
				storePath = fullPath + ".gz"
			}
		}
	}

	if !compressed {
		dataToStore = content
	}

	// Update metadata with compression info
	if metadata != nil {
		if metadata.Custom == nil {
			metadata.Custom = make(map[string]string)
		}
		if compressed {
			metadata.Custom["compressed"] = "true"
			metadata.Custom["original_size"] = fmt.Sprintf("%d", len(content))
			metadata.CompressedSize = int64(len(dataToStore))
		}
	}

	// Write content file
	if err := os.WriteFile(storePath, dataToStore, 0644); err != nil {
		return fmt.Errorf("failed to write file %s: %w", storePath, err)
	}

	return nil
}

// Get retrieves content from the given key
func (s *LocalStorage) Get(ctx context.Context, key string) ([]byte, error) {
	fullPath := s.keyToPath(key)

	// Check if .gz version exists (compressed file)
	gzPath := fullPath + ".gz"
	if _, err := os.Stat(gzPath); err == nil {
		content, err := os.ReadFile(gzPath)
		if err != nil {
			return nil, fmt.Errorf("failed to read compressed file %s: %w", gzPath, err)
		}
		decompressed, err := DecompressGzip(content)
		if err != nil {
			return nil, fmt.Errorf("decompression failed for %s (file may be corrupted): %w", key, err)
		}
		return decompressed, nil
	}

	// Fall back to uncompressed file
	content, err := os.ReadFile(fullPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("file not found: %s", key)
		}
		return nil, fmt.Errorf("failed to read file %s: %w", fullPath, err)
	}

	return content, nil
}

// GetInfo retrieves file information without content
func (s *LocalStorage) GetInfo(ctx context.Context, key string) (*FileInfo, error) {
	fullPath := s.keyToPath(key)

	// Check for compressed version first
	actualPath := fullPath
	isCompressed := false
	gzPath := fullPath + ".gz"
	if _, err := os.Stat(gzPath); err == nil {
		actualPath = gzPath
		isCompressed = true
	}

	stat, err := os.Stat(actualPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, fmt.Errorf("file not found: %s", key)
		}
		return nil, fmt.Errorf("failed to stat file %s: %w", actualPath, err)
	}

	// Compute checksum
	checksum, err := s.computeFileChecksum(actualPath)
	if err != nil {
		return nil, fmt.Errorf("failed to compute checksum: %w", err)
	}

	info := &FileInfo{
		Key:        key,
		Size:       stat.Size(),
		Checksum:   checksum,
		ModifiedAt: stat.ModTime(),
	}

	// Add compression info if applicable
	if isCompressed {
		info.Metadata = &Metadata{
			Custom: map[string]string{"compressed": "true"},
		}
	}

	return info, nil
}

// Exists checks if a file exists at the given key
func (s *LocalStorage) Exists(ctx context.Context, key string) (bool, error) {
	fullPath := s.keyToPath(key)

	// Check for uncompressed file
	if _, err := os.Stat(fullPath); err == nil {
		return true, nil
	}

	// Check for compressed file
	gzPath := fullPath + ".gz"
	if _, err := os.Stat(gzPath); err == nil {
		return true, nil
	}

	return false, nil
}

// Delete removes a file at the given key
func (s *LocalStorage) Delete(ctx context.Context, key string) error {
	fullPath := s.keyToPath(key)

	// Delete uncompressed content file
	if err := os.Remove(fullPath); err != nil && !os.IsNotExist(err) {
		return fmt.Errorf("failed to delete file %s: %w", fullPath, err)
	}

	// Delete compressed file if exists
	gzPath := fullPath + ".gz"
	if err := os.Remove(gzPath); err != nil && !os.IsNotExist(err) {
		// Ignore deletion errors for gz file
	}

	return nil
}

// List returns all keys matching the given prefix
func (s *LocalStorage) List(ctx context.Context, prefix string) ([]string, error) {
	searchPath := s.keyToPath(prefix)

	// Ensure search path is a directory or get parent directory
	stat, err := os.Stat(searchPath)
	if err != nil {
		if os.IsNotExist(err) {
			// Get parent directory
			searchPath = filepath.Dir(searchPath)
			if _, err := os.Stat(searchPath); os.IsNotExist(err) {
				return []string{}, nil
			}
		} else {
			return nil, fmt.Errorf("failed to stat path %s: %w", searchPath, err)
		}
	} else if !stat.IsDir() {
		searchPath = filepath.Dir(searchPath)
	}

	var keys []string
	err = filepath.Walk(searchPath, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return err
		}

		// Skip directories
		if info.IsDir() {
			return nil
		}

		// Convert path back to key, stripping .gz extension if present
		key := s.pathToKey(path)
		key = strings.TrimSuffix(key, ".gz")

		// Filter by prefix
		if strings.HasPrefix(key, prefix) {
			keys = append(keys, key)
		}

		return nil
	})

	if err != nil {
		return nil, fmt.Errorf("failed to list files: %w", err)
	}

	return keys, nil
}

// GetChecksum returns the checksum for a file
func (s *LocalStorage) GetChecksum(ctx context.Context, key string) (string, error) {
	fullPath := s.keyToPath(key)

	// Check for compressed version first
	gzPath := fullPath + ".gz"
	if _, err := os.Stat(gzPath); err == nil {
		return s.computeFileChecksum(gzPath)
	}

	return s.computeFileChecksum(fullPath)
}

// keyToPath converts a storage key to a filesystem path
func (s *LocalStorage) keyToPath(key string) string {
	// Clean the key to prevent path traversal
	cleanKey := filepath.Clean(key)
	cleanKey = strings.TrimPrefix(cleanKey, "/")
	cleanKey = strings.TrimPrefix(cleanKey, "\\")

	return filepath.Join(s.basePath, cleanKey)
}

// pathToKey converts a filesystem path to a storage key
func (s *LocalStorage) pathToKey(path string) string {
	relPath, err := filepath.Rel(s.basePath, path)
	if err != nil {
		return path
	}
	// Normalize to forward slashes for consistency
	return strings.ReplaceAll(relPath, "\\", "/")
}

// computeFileChecksum computes SHA256 checksum for a file
func (s *LocalStorage) computeFileChecksum(path string) (string, error) {
	file, err := os.Open(path)
	if err != nil {
		return "", fmt.Errorf("failed to open file: %w", err)
	}
	defer file.Close()

	hash := sha256.New()
	if _, err := io.Copy(hash, file); err != nil {
		return "", fmt.Errorf("failed to compute hash: %w", err)
	}

	return hex.EncodeToString(hash.Sum(nil)), nil
}

// ComputeChecksum computes SHA256 checksum for content
func ComputeChecksum(content []byte) string {
	hash := sha256.Sum256(content)
	return hex.EncodeToString(hash[:])
}

// GetBasePath returns the base path for this storage
func (s *LocalStorage) GetBasePath() string {
	return s.basePath
}

// BuildArchiveKey builds a storage key for an archive file
func BuildArchiveKey(chainSlug string, date time.Time, filename string) string {
	dateStr := date.Format("2006-01-02")
	return fmt.Sprintf("archives/%s/%s/%s", chainSlug, dateStr, filename)
}

// BuildExpandedKey builds a storage key for an expanded file from a ZIP
func BuildExpandedKey(chainSlug string, date time.Time, parentFilename, innerFilename string) string {
	dateStr := date.Format("2006-01-02")
	// Remove .zip extension from parent
	parentBase := strings.TrimSuffix(parentFilename, ".zip")
	parentBase = strings.TrimSuffix(parentBase, ".ZIP")
	return fmt.Sprintf("expanded/%s/%s/%s/%s", chainSlug, dateStr, parentBase, innerFilename)
}

// ============================================================================
// IntermediateStorage Implementation
// ============================================================================

// WriteIntermediateJSON writes a JSON object to intermediate storage.
func (s *LocalStorage) WriteIntermediateJSON(ctx context.Context, runID, filename string, data interface{}) error {
	key := BuildIntermediateKey(runID, filename)

	jsonBytes, err := json.Marshal(data)
	if err != nil {
		return fmt.Errorf("failed to marshal JSON: %w", err)
	}

	return s.Put(ctx, key, jsonBytes, &Metadata{
		ContentType: "application/json",
		Custom: map[string]string{
			"file_type": "json",
		},
	})
}

// ReadIntermediateJSON reads a JSON object from intermediate storage.
func (s *LocalStorage) ReadIntermediateJSON(ctx context.Context, runID, filename string, dest interface{}) error {
	key := BuildIntermediateKey(runID, filename)

	content, err := s.Get(ctx, key)
	if err != nil {
		return fmt.Errorf("failed to read intermediate file %s: %w", key, err)
	}

	if err := json.Unmarshal(content, dest); err != nil {
		return fmt.Errorf("failed to unmarshal JSON from %s: %w", key, err)
	}

	return nil
}

// ListIntermediateFiles returns all files in intermediate storage for a run.
func (s *LocalStorage) ListIntermediateFiles(ctx context.Context, runID string) ([]string, error) {
	prefix := "intermediate/" + runID + "/"

	keys, err := s.List(ctx, prefix)
	if err != nil {
		return nil, fmt.Errorf("failed to list intermediate files for run %s: %w", runID, err)
	}

	// Return just the filenames, not full keys
	filenames := make([]string, 0, len(keys))
	for _, key := range keys {
		filename := strings.TrimPrefix(key, prefix)
		if filename != "" {
			filenames = append(filenames, filename)
		}
	}

	return filenames, nil
}

// DeleteIntermediateDir deletes all intermediate files for a run.
func (s *LocalStorage) DeleteIntermediateDir(ctx context.Context, runID string) error {
	prefix := "intermediate/" + runID + "/"

	keys, err := s.List(ctx, prefix)
	if err != nil {
		return fmt.Errorf("failed to list intermediate files for deletion: %w", err)
	}

	for _, key := range keys {
		if err := s.Delete(ctx, key); err != nil {
			return fmt.Errorf("failed to delete intermediate file %s: %w", key, err)
		}
	}

	// Also try to remove the directory itself
	dirPath := s.keyToPath(prefix)
	os.RemoveAll(dirPath) // Ignore errors - directory might not be empty or might not exist

	return nil
}

// BuildStoreDataKey builds a key for parsed store data.
func BuildStoreDataKey(runID, storeIdentifier string) string {
	// Sanitize store identifier for use in filename
	safeIdentifier := strings.ReplaceAll(storeIdentifier, "/", "_")
	safeIdentifier = strings.ReplaceAll(safeIdentifier, "\\", "_")
	safeIdentifier = strings.ReplaceAll(safeIdentifier, ":", "_")
	return "stores/" + safeIdentifier + ".json"
}

// BuildManifestKey builds a key for the run manifest.
func BuildManifestKey() string {
	return "manifest.json"
}

File: src/lib/storage/local.ts
import { createHash } from "node:crypto";
import { createReadStream, createWriteStream } from "node:fs";
import {
	mkdir,
	readdir,
	readFile,
	rm,
	stat,
	writeFile,
} from "node:fs/promises";
import * as path from "node:path";
import { Readable } from "node:stream";
import { pipeline } from "node:stream/promises";
import { createGunzip, createGzip } from "node:zlib";
import {
	compressGzip,
	decompressGzip,
	shouldCompressSmart,
} from "./compression";
import type { FileInfo, Storage, StorageMetadata } from "./index";

/** Minimum size in bytes before compression is attempted */
export const MIN_COMPRESSION_SIZE = 1024; // 1KB

/** Maximum decompressed file size (500MB) to prevent memory exhaustion */
const MAX_DECOMPRESSED_SIZE = 500 * 1024 * 1024;

/**
 * LocalStorage implements the Storage interface using local filesystem.
 * Matches Go implementation in services/price-service/internal/storage/local.go
 */
export class LocalStorage implements Storage {
	constructor(private basePath: string) {}

	/**
	 * Store content at the given key with optional metadata.
	 * Compresses data if applicable based on file type and size.
	 */
	async put(
		key: string,
		data: Buffer,
		metadata?: StorageMetadata,
	): Promise<void> {
		const fullPath = this.keyToPath(key);

		// Ensure parent directory exists
		const dir = path.dirname(fullPath);
		await mkdir(dir, { recursive: true });

		let dataToStore = data;
		let compressed = false;
		let storePath = fullPath;

		// Determine if we should compress using smart compression
		// This checks both file type and if file is already compressed
		if (data.length >= MIN_COMPRESSION_SIZE) {
			const filename = metadata?.originalName || path.basename(key);
			const fileType = metadata?.custom?.file_type;

			if (shouldCompressSmart(filename, fileType)) {
				const compressedData = await compressGzip(data);
				// Only use compression if it actually reduces size
				if (compressedData.length < data.length) {
					dataToStore = compressedData;
					compressed = true;
					storePath = `${fullPath}.gz`;
				}
			}
		}

		// Update metadata with compression info
		if (metadata) {
			if (!metadata.custom) {
				metadata.custom = {};
			}
			if (compressed) {
				metadata.custom.compressed = "true";
				metadata.custom.original_size = String(data.length);
				metadata.compressedSize = dataToStore.length;
			}
		}

		// Write content file
		await writeFile(storePath, dataToStore);

		// Write metadata sidecar file if metadata is provided
		if (metadata) {
			const metaPath = `${fullPath}.meta.json`;
			await writeFile(metaPath, JSON.stringify(metadata));
		}
	}

	/**
	 * Store content using streaming compression directly to file.
	 * More memory-efficient for large files.
	 */
	async putStreaming(
		key: string,
		data: Buffer,
		metadata?: StorageMetadata,
	): Promise<void> {
		const fullPath = this.keyToPath(key);

		// Ensure parent directory exists
		const dir = path.dirname(fullPath);
		await mkdir(dir, { recursive: true });

		// Determine if we should compress using smart compression
		const filename = metadata?.originalName || path.basename(key);
		const fileType = metadata?.custom?.file_type;
		const shouldCompressFile =
			data.length >= MIN_COMPRESSION_SIZE &&
			shouldCompressSmart(filename, fileType);

		if (shouldCompressFile) {
			const storePath = `${fullPath}.gz`;
			const readStream = Readable.from(data);
			const writeStream = createWriteStream(storePath);
			await pipeline(readStream, createGzip(), writeStream);

			// Update metadata
			if (metadata) {
				if (!metadata.custom) {
					metadata.custom = {};
				}
				metadata.custom.compressed = "true";
				metadata.custom.original_size = String(data.length);
				const stats = await stat(storePath);
				metadata.compressedSize = stats.size;
			}
		} else {
			await writeFile(fullPath, data);
		}
	}

	/**
	 * Retrieve content from the given key.
	 * Automatically decompresses if stored as .gz file.
	 */
	async get(key: string): Promise<Buffer> {
		const fullPath = this.keyToPath(key);

		// Check if .gz version exists (compressed file)
		const gzPath = `${fullPath}.gz`;
		try {
			const stats = await stat(gzPath);
			if (stats.isFile()) {
				const content = await readFile(gzPath);
				return decompressGzip(content);
			}
		} catch {
			// .gz file doesn't exist, try uncompressed
		}

		// Fall back to uncompressed file
		try {
			return await readFile(fullPath);
		} catch (err) {
			if ((err as NodeJS.ErrnoException).code === "ENOENT") {
				throw new Error(`file not found: ${key}`);
			}
			throw err;
		}
	}

	/**
	 * Retrieve content using streaming decompression.
	 * More memory-efficient for large files.
	 */
	async getStreaming(key: string): Promise<Buffer> {
		const fullPath = this.keyToPath(key);

		// Check if .gz version exists (compressed file)
		const gzPath = `${fullPath}.gz`;
		try {
			const stats = await stat(gzPath);
			if (stats.isFile()) {
				// Safety check: don't decompress if compressed size is unreasonably large
				// A 500MB compressed file could decompress to several GB
				if (stats.size > MAX_DECOMPRESSED_SIZE) {
					throw new Error(
						`Compressed file too large to decompress: ${stats.size} bytes (max ${MAX_DECOMPRESSED_SIZE})`,
					);
				}

				let totalSize = 0;
				const chunks: Buffer[] = [];
				const readStream = createReadStream(gzPath);
				await pipeline(readStream, createGunzip(), async (source) => {
					for await (const chunk of source) {
						totalSize += (chunk as Buffer).length;
						if (totalSize > MAX_DECOMPRESSED_SIZE) {
							throw new Error(
								`Decompressed file exceeds maximum size: ${totalSize} bytes (max ${MAX_DECOMPRESSED_SIZE})`,
							);
						}
						chunks.push(chunk as Buffer);
					}
				});
				return Buffer.concat(chunks);
			}
		} catch {
			// .gz file doesn't exist, try uncompressed
		}

		// Fall back to uncompressed file
		try {
			return await readFile(fullPath);
		} catch (err) {
			if ((err as NodeJS.ErrnoException).code === "ENOENT") {
				throw new Error(`file not found: ${key}`);
			}
			throw err;
		}
	}

	/**
	 * Retrieve file information without content.
	 */
	async getInfo(key: string): Promise<FileInfo> {
		const fullPath = this.keyToPath(key);

		// Check for compressed version first
		let actualPath = fullPath;
		let isCompressed = false;
		const gzPath = `${fullPath}.gz`;

		try {
			const gzStats = await stat(gzPath);
			if (gzStats.isFile()) {
				actualPath = gzPath;
				isCompressed = true;
			}
		} catch {
			// .gz file doesn't exist
		}

		try {
			const stats = await stat(actualPath);
			const checksum = await this.computeFileChecksum(actualPath);

			const info: FileInfo = {
				key,
				size: stats.size,
				checksum,
				modifiedAt: stats.mtime,
			};

			// Read metadata from sidecar file if it exists
			const metaPath = `${fullPath}.meta.json`;
			try {
				const metaContent = await readFile(metaPath, "utf-8");
				info.metadata = JSON.parse(metaContent) as StorageMetadata;
			} catch {
				// No metadata file, add compression info if applicable
				if (isCompressed) {
					info.metadata = {
						custom: { compressed: "true" },
					};
				}
			}

			return info;
		} catch (err) {
			if ((err as NodeJS.ErrnoException).code === "ENOENT") {
				throw new Error(`file not found: ${key}`);
			}
			throw err;
		}
	}

	/**
	 * Check if a file exists at the given key.
	 * Checks both uncompressed and .gz versions.
	 */
	async exists(key: string): Promise<boolean> {
		const fullPath = this.keyToPath(key);

		// Check for uncompressed file
		try {
			const stats = await stat(fullPath);
			if (stats.isFile()) {
				return true;
			}
		} catch {
			// File doesn't exist
		}

		// Check for compressed file
		const gzPath = `${fullPath}.gz`;
		try {
			const stats = await stat(gzPath);
			if (stats.isFile()) {
				return true;
			}
		} catch {
			// .gz file doesn't exist
		}

		return false;
	}

	/**
	 * Delete a file at the given key.
	 * Removes both uncompressed and .gz versions if they exist, plus metadata sidecar.
	 */
	async delete(key: string): Promise<void> {
		const fullPath = this.keyToPath(key);

		// Delete uncompressed content file
		try {
			await rm(fullPath);
		} catch (err) {
			if ((err as NodeJS.ErrnoException).code !== "ENOENT") {
				throw err;
			}
		}

		// Delete compressed file if exists
		const gzPath = `${fullPath}.gz`;
		try {
			await rm(gzPath);
		} catch {
			// Ignore deletion errors for gz file
		}

		// Delete metadata sidecar file if exists
		const metaPath = `${fullPath}.meta.json`;
		try {
			await rm(metaPath);
		} catch {
			// Ignore deletion errors for metadata file
		}
	}

	/**
	 * List all keys matching the given prefix.
	 * Strips .gz suffix from compressed files.
	 */
	async list(prefix: string): Promise<string[]> {
		const searchPath = this.keyToPath(prefix);

		// Ensure search path exists or get parent directory
		let basePath = searchPath;
		try {
			const stats = await stat(searchPath);
			if (!stats.isDirectory()) {
				basePath = path.dirname(searchPath);
			}
		} catch {
			basePath = path.dirname(searchPath);
			try {
				await stat(basePath);
			} catch {
				return [];
			}
		}

		const keys: string[] = [];
		const seen = new Set<string>();

		await this.walkDir(basePath, async (filePath) => {
			// Convert path back to key, stripping .gz extension if present
			let key = this.pathToKey(filePath);
			if (key.endsWith(".gz")) {
				key = key.slice(0, -3);
			}

			// Filter by prefix and avoid duplicates
			if (key.startsWith(prefix) && !seen.has(key)) {
				seen.add(key);
				keys.push(key);
			}
		});

		return keys;
	}

	/**
	 * Get the checksum for a file.
	 * Returns SHA256 hash of the stored file (compressed or uncompressed).
	 */
	async getChecksum(key: string): Promise<string> {
		const fullPath = this.keyToPath(key);

		// Check for compressed version first
		const gzPath = `${fullPath}.gz`;
		try {
			const stats = await stat(gzPath);
			if (stats.isFile()) {
				return this.computeFileChecksum(gzPath);
			}
		} catch {
			// .gz file doesn't exist
		}

		return this.computeFileChecksum(fullPath);
	}

	/**
	 * Get the base path for this storage.
	 */
	getBasePath(): string {
		return this.basePath;
	}

	/**
	 * Convert a storage key to a filesystem path.
	 * Cleans the key to prevent path traversal attacks.
	 */
	private keyToPath(key: string): string {
		// Normalize path separators and collapse redundant segments.
		let cleanKey = key.replace(/\\/g, "/");
		cleanKey = path.posix.normalize(cleanKey);
		cleanKey = cleanKey.replace(/^[/]+/, "");

		// Reject parent-directory traversal segments but allow filenames with "..".
		const parts = cleanKey.split("/").filter(Boolean);
		if (parts.some((part) => part === "..")) {
			throw new Error(`invalid key: path traversal detected in "${key}"`);
		}

		return path.join(this.basePath, cleanKey);
	}

	/**
	 * Convert a filesystem path to a storage key.
	 */
	private pathToKey(filePath: string): string {
		const relPath = path.relative(this.basePath, filePath);
		// Normalize to forward slashes for consistency
		return relPath.replace(/\\/g, "/");
	}

	/**
	 * Compute SHA256 checksum for a file using streaming.
	 */
	private async computeFileChecksum(filePath: string): Promise<string> {
		return new Promise((resolve, reject) => {
			const hash = createHash("sha256");
			const stream = createReadStream(filePath);

			stream.on("data", (chunk) => hash.update(chunk));
			stream.on("end", () => resolve(hash.digest("hex")));
			stream.on("error", reject);
		});
	}

	/**
	 * Recursively walk a directory and call callback for each file.
	 */
	private async walkDir(
		dir: string,
		callback: (filePath: string) => Promise<void>,
	): Promise<void> {
		try {
			const entries = await readdir(dir, { withFileTypes: true });

			for (const entry of entries) {
				const fullPath = path.join(dir, entry.name);
				if (entry.isDirectory()) {
					await this.walkDir(fullPath, callback);
				} else if (entry.isFile()) {
					await callback(fullPath);
				}
			}
		} catch {
			// Directory doesn't exist or not accessible
		}
	}
}

/**
 * Compute SHA256 checksum for content.
 */
export function computeChecksum(content: Buffer): string {
	const hash = createHash("sha256");
	hash.update(content);
	return hash.digest("hex");
}
